### PR TITLE
Support libftdi1 too in the ftdidmx plugin

### PR DESCRIPTION
**Pull request by Keeper-of-the-Keys**
libftdi1 support

### CODE REVIEW

**Comment by peternewman** [comment]
Well my build progresses a lot further ( https://travis-ci.org/peternewman/ola/builds/384539453 ), so it's something different here: https://github.com/peternewman/ola/compare/0.10-libftdi1...Keeper-of-the-Keys:0.10-libftdi1

**Comment by peternewman** [comment]
And no sooner had I said it, I spotted the error. Fixed here https://github.com/OpenLightingProject/ola/pull/1422/commits/c8d17420b0044a07162db2b438e4ba2ac89c86b6 , it's just building now.

File: olad/Makefile.mk [hunk 1]
```diff
@@ -72,10 +72,6 @@ olad_olad_LDADD = olad/libolaserver.la \
                   common/libolacommon.la \
                   ola/libola.la
 
-if USE_FTDI
-olad_olad_LDADD += -lftdi -lusb
```
Comment (peternewman) [inline olad/Makefile.mk]: You don't seem to have put these lines back in anywhere else in the plugin.

File: plugins/ftdidmx/FtdiWidget.cpp [hunk 2]
```diff
@@ -127,7 +136,11 @@ void FtdiWidget::Widgets(vector<FtdiWidgetInfo> *widgets) {
       ftdi_device_list* current_device = list;
 
       while (current_device != NULL) {
+#if HAVE_LIBFTDI1
+        struct libusb_device *dev = current_device->dev;
+#else
         struct usb_device *dev = current_device->dev;
+#endif //HAVE_LIBFTDI1
```
Comment (peternewman) [inline plugins/ftdidmx/FtdiWidget.cpp]: Can you fix the lint error here:
https://travis-ci.org/OpenLightingProject/ola/jobs/384567230#L1754

Comment (Keeper-of-the-Keys) [reply plugins/ftdidmx/FtdiWidget.cpp]: Fixed

File: olad/Makefile.mk (hunk 1 above)
Comment (peternewman) [reply olad/Makefile.mk]: Now done.

File: plugins/ftdidmx/Makefile.mk [hunk 3]
```diff
@@ -13,7 +13,14 @@ plugins_ftdidmx_libolaftdidmx_la_SOURCES = \
...
 plugins_ftdidmx_libolaftdidmx_la_LIBADD = \
-    $(libftdi_LIBS) \
     common/libolacommon.la \
     olad/plugin_api/libolaserverplugininterface.la
+if HAVE_LIBFTDI1
+plugins_ftdidmx_libolaftdidmx_la_LIBADD += \
+    $(libftdi1_LIBS)
+else
+plugins_ftdidmx_libolaftdidmx_la_LIBADD += \
+    $(libftdi_LIBS)
```
Comment (peternewman) [inline plugins/ftdidmx/Makefile.mk]: I think this wants to be a zero now given I/we changed this line to be libftdi0 from libftdi:
https://github.com/OpenLightingProject/ola/pull/1422/files#diff-67e997bcfdac55191033d57a16d1408aR624

Comment (Keeper-of-the-Keys) [reply plugins/ftdidmx/Makefile.mk]: Fixed

File: configure.ac [hunk 4]
```diff
@@ -612,19 +612,33 @@ AC_ARG_ENABLE(
...
   [AS_HELP_STRING([--enable-ja-rule], [Build the Ja Rule test tool])])
 
-# libftdi: this requires libusb-0.1.4+ don't know if I should also
+# libftdi0: this requires libusb-0.1.4+ don't know if I should also
 # check for it, since the libusb check above is not for 0.1.4 but for libusb-1
 AC_ARG_ENABLE(
   [libftdi],
   [AS_HELP_STRING([--disable-libftdi],
-                  [Avoid using libftdi even if it exists])])
+                  [Avoid using libftdi or libftdi1 even if either exists])])
```
Comment (peternewman) [inline configure.ac]: This should probably say libftdi0 or libftdi1 for clarity.

File: configure.ac [hunk 5]
```diff
@@ -612,19 +612,33 @@ AC_ARG_ENABLE(
...
+AM_CONDITIONAL([HAVE_LIBFTDI1], [test "x$have_libftdi1" = xyes])
+AS_IF([test "x$have_libftdi1" = xyes],
+      [AC_DEFINE([HAVE_LIBFTDI1], [1], [define if libftdi1 is installed])])
+have_libftdi="no"
+if test "x$have_libftdi0" = xyes || test "x$have_libftdi1" = xyes; then
+  have_libftdi="yes"
+fi
 AS_IF([test "x$have_libftdi" = xyes],
-      [AC_DEFINE([HAVE_LIBFTDI], [1], [define if libftdi is installed])])
+      [AC_DEFINE([HAVE_LIBFTDI], [1], [define if libftdi or libftdi1 is installed])])
```
Comment (peternewman) [inline configure.ac]: Likewise here.

File: plugins/ftdidmx/Makefile.mk [hunk 6]
```diff
@@ -13,7 +13,14 @@ plugins_ftdidmx_libolaftdidmx_la_SOURCES = \
     plugins/ftdidmx/FtdiWidget.cpp \
     plugins/ftdidmx/FtdiWidget.h
 plugins_ftdidmx_libolaftdidmx_la_LIBADD = \
-    $(libftdi_LIBS) \
     common/libolacommon.la \
     olad/plugin_api/libolaserverplugininterface.la
+if HAVE_LIBFTDI1
+plugins_ftdidmx_libolaftdidmx_la_LIBADD += \
```
Comment (peternewman) [inline plugins/ftdidmx/Makefile.mk]: This is narrow enough it could all be one one line.

File: plugins/ftdidmx/Makefile.mk [hunk 7]
```diff
@@ -13,7 +13,14 @@ plugins_ftdidmx_libolaftdidmx_la_SOURCES = \
...
+    $(libftdi1_LIBS)
+else
+plugins_ftdidmx_libolaftdidmx_la_LIBADD += \
+    $(libftdi0_LIBS)
```
Comment (peternewman) [inline plugins/ftdidmx/Makefile.mk]: And again

File: .travis.yml [hunk 8]
```diff
@@ -45,24 +45,42 @@ matrix:
...
+      - LIBFTDI='0'
     - os: osx
       osx_image: xcode9.3
       compiler: gcc
       env:
       - TASK='compile'
       - CPPUNIT='1.14'
+      - LIBFTDI='0'
+    - os: osx
+      osx_image: xcode9.3beta
```
Comment (peternewman) [inline .travis.yml]: This just wants to be xcode9.3 now please.

File: .travis.yml [hunk 9]
```diff
@@ -45,24 +45,42 @@ matrix:
...
+      compiler: clang
+      env:
+      - TASK='compile'
+      - CPPUNIT='1.14'
+      - LIBFTDI='1'
+    - os: osx
+      osx_image: xcode9.3beta
```
Comment (peternewman) [inline .travis.yml]: Again

File: plugins/ftdidmx/FtdiWidget.cpp [hunk 10]
```diff
@@ -32,8 +32,17 @@
...
+#include <config.h>
+#endif  // HAVE_CONFIG_H
+
+#ifdef HAVE_LIBFTDI1
+#include <libftdi1/ftdi.h>
+#else
 #include <ftdi.h>
+#endif  // HAVE_LIBFTDI1
+
+#include <strings.h>
```
Comment (peternewman) [inline plugins/ftdidmx/FtdiWidget.cpp]: Can you sort these two .h includes alphabetically please?

Comment (peternewman) [reply plugins/ftdidmx/FtdiWidget.cpp]: I've done this now @Keeper-of-the-Keys

**Comment by peternewman** [comment]
CI blocked behind #1432 .

**Comment by peternewman** [comment]
No worries @Keeper-of-the-Keys . Can you just confirm my queries, then we'll hit merge (also as before feel free to add yourself to AUTHORS if you want):

One question looking at the changes made @Keeper-of-the-Keys . Can you try compiling without libusb (i.e. add --disable-usbdmx to configure args) given the comments here:
https://github.com/OpenLightingProject/ola/pull/1422/files#diff-67e997bcfdac55191033d57a16d1408aR615

Given we also used to add -lusb:
https://github.com/OpenLightingProject/ola/pull/1422/files#diff-f03ba74a5a591aa49368e0431358829eL76

I suspect it may fail (especially with libftdi1), in which case we might need to tweak configure a bit.

Also I assume you've tested the latest iteration with a dongle of some kind?

**Comment by Keeper-of-the-Keys** [comment]
It compiled fine with --disable-usbdmx and runs OK too as far as I can tell.

Added myself to Authors as requested, what is this certificate error that Travis is/was complaining about?

**Comment by Keeper-of-the-Keys** [comment]
@peternewman Note that I don't have any old systems to run on so I only test libftdi1

**Comment by peternewman** [comment]
I guess the -lusb was a legacy thing then.

Thanks. I'll try and test libftdi0 at some point, but given the nature of the changes, I'm fairly happy that given it still compiles it should be fine. As soon as #1432 gets merged I'll merge this.

What was the certificate error you were seeing?

Yes, I think AllowMultiPortPatching() and multiport naming is next after that. I'd be happy for AllowMultiPortPatching() to go into the 0.10 branch, if you've got a multiport device to test it with. I think the naming stuff probably ought to go into master for the next release though, as it's more feature than bug.

**Comment by peternewman** [comment]
I've just tested libftdi1 and libftdi0 on Ubuntu 16.04 successfully.

**Pull request merged**